[Multilib] Custom flags YAML parsing

vhscampos opened the pull request:
This patch is the first step to extend the current multilib system to support the selection of library variants which do not correspond to existing command-line options.

Proposal can be found in https://discourse.llvm.org/t/rfc-multilib-custom-flags/81058

The multilib mechanism supports libraries that target code generation or language options such as `--target`, `-mcpu`, `-mfpu`, `-mbranch-protection`. However, some library variants are particular to features that do not correspond to any command-line options. Examples include variants for multithreading and semihosting. 

This work introduces a way to instruct the multilib system to consider these features in library selection. This particular patch comprises a new section in `multilib.yaml` to declare flags for which no option exists. Henceforth this sort of flag will be called `custom flag` for clarity.

The `multilib.yaml` file will have a new section called Flags which contains the declarations of the target’s custom flags:

```yaml
Flags:
- Name: multithreaded
  Values:
  - Name: no-multithreaded
    MacroDefines: [__SINGLE_THREAD__]
  - Name: multithreaded
  Default: no-multithreaded

- Name: io
  Values:
    - Name: io-none
    - Name: io-semihosting
      MacroDefines: [SEMIHOSTING]
    - Name: io-linux-syscalls
      MacroDefines: [LINUX_SYSCALLS, HOSTED=1]
   Default: io-none
```
- Name: the name to categorize a flag.
- Values: a list of possible values.
- Default: it specifies which value this flag should take if not specified in the command-line invocation. It must be one value from the Values field.

Each flag Value follows this description:
- Name (required): the name of the custom flag value (string). This is the string to be used in `-fmultilib-flag=<string>`.
- MacroDefines (optional): a list of strings to be used as macro definitions. Each string
  is fed into the driver as ``-D<string>``.

A Default value is useful to save users from specifying custom flags that have a most commonly used value.

The namespace of flag values is common across all flags. This means that flag values must be unique.

vhscampos committed:
[Multilib] Custom flags YAML parsing

This patch adds support for custom flags in the multilib YAML file.

Details about this change can be found in:
https://discourse.llvm.org/t/rfc-multilib-custom-flags/81058

CustomFlagDeclaration objects are instantiated using shared_ptr. This is
motivated by the fact that each custom flag value,
CustomFlagValueDetail, contains a back reference to their corresponding
flag declaration.

Since the CustomFlagDeclaration objects are transferred from the YAML
parser to the MultilibSet instance after the parsing is finished, back
references implemented as raw pointers would become dangling. This would
need to be remediated in the copy/move constructors by updating the
pointer.

Therefore it's just simpler and less error-prone to have all references
to CustomFlagDeclaration, including the back reference, as shared_ptr.
This way dangling pointers are not a concern.

vhscampos commented on clang/include/clang/Driver/Multilib.h:
```diff
@@ -101,6 +101,25 @@ class Multilib {
 raw_ostream &operator<<(raw_ostream &OS, const Multilib &M);
 
+namespace custom_flag {
+struct CustomFlagDeclaration;
```
Fixed

vhscampos commented on clang/lib/Driver/Multilib.cpp:
```diff
@@ -259,11 +266,69 @@ template <> struct llvm::yaml::MappingTraits<MultilibSet::FlagMatcher> {
+          return !Value.Decl || Value.Decl != V;
+        }))
+      return "custom flag value missing reference to its custom flag "
+             "declaration";
```
Fixed

vhscampos commented on clang/include/clang/Driver/Multilib.h:
```diff
@@ -101,6 +101,25 @@ class Multilib {
+struct CustomFlagDeclaration {
+  std::string Name;
+  SmallVector<CustomFlagValueDetail> ValueList;
+  size_t DefaultValueIdx = ~0UL;
```
Fixed

smithp35 commented:
> This patch adds support for custom flags in the multilib YAML file.
> 
> Details about this change can be found in this thread: https://discourse.llvm.org/t/rfc-multilib-custom-flags/81058

Apologies in advance for being a pain. Prior to commit, assuming that the RFC has stabilised, please could the description be updated to give a summary of custom flags and what this part of the patch adds. It is fine to link to the RFC, although that may not be as permanent as we'd like and it makes it harder for someone searching for the commit log to find the patch.

vhscampos commented:
Thanks @smithp35. Before any of the PRs is merged, I will update the descriptions to make them more self-contained.

Furthermore, `Multilib.rst` must indeed be updated. Will do it.

vhscampos commented:
Created PR to amend `Multilib.rst`: https://github.com/llvm/llvm-project/pull/114998

vhscampos committed:
Change in specification: DriverArgs now MacroDefines

We've decided to restrict the driver arguments that can be specified to
only macro definitions. Hence this patch adjusts the YAML parsing for
that.

lenary reviewed (approved):
LGTM

vhscampos commented:
I've reverted this patch because apparently Clang driver does not meld well with `std::shared_ptr`: memory leaks due to custom allocators I think. Thus I've redone the part to use raw pointers instead: https://github.com/llvm/llvm-project/pull/122903
